Make upload progress observable

jplatte opened the pull request:
This works, it just comes with a very ugly change to the `HttpSend` trait that I feel like is making it no longer suiteable as public API.

I think we should instead settle on reqwest, and make the reqwest client configurable for users.

jplatte commented on crates/matrix-sdk/src/http_client.rs:
```diff
@@ -455,9 +493,23 @@ impl HttpSend for reqwest::Client {
     ) -> Result<http::Response<Bytes>, HttpError> {
         #[allow(unused_mut)]
-        let mut request = reqwest::Request::try_from(request)?;
+        let mut request = if send_progress.subscriber_count() != 0 {
+            send_progress.update(|p| p.total += request.body().len());
+            reqwest::Request::try_from(request.map(|body| {
+                let chunks = stream::iter(BytesChunks::new(body, 8192).map(
+                    move |chunk| -> Result<_, Infallible> {
+                        send_progress.update(|p| p.current += chunk.len());
+                        Ok(chunk)
+                    },
+                ));
+                reqwest::Body::wrap_stream(chunks)
+            }))?
+        } else {
+            reqwest::Request::try_from(request)?
+        };
```
@poljar this is what's required for tracking upload progress with reqwest. I think anybody implementing their own `HttpSend` would have a really hard time figuring out the equivalent solution for whatever library they are using, and default to not tracking progress at all. Do you think we could drop support for different HTTP transports, and instead allow users to configure the `reqwest::Client` instance used by the SDK client?

poljar replied:
It sounds a bit scary to remove it, but I see your point, things becomes quite library specific.

As far as I know, the only requirement we have is that people should be able to customize the HTTP headers each request sends out, so we should be able to do this. I guess if we do end up needing it we can come back and think of a better abstraction for this.

Hywan reviewed (approved):
Good job!

poljar commented on crates/matrix-sdk/src/http_client.rs:
```diff
@@ -83,6 +91,7 @@ pub trait HttpSend: AsyncTraitDeps {
     ///         &self,
     ///         request: http::Request<Bytes>,
     ///         timeout: Duration,
+    ///         _send_progress: SharedObservable<TransmissionProgress>,
```
```suggestion
    ///         send_progress: SharedObservable<TransmissionProgress>,
```

poljar commented on crates/matrix-sdk-ui/src/timeline/futures.rs:
```diff
@@ -0,0 +1,68 @@
+                .file_name()
+                .ok_or(Error::InvalidAttachmentFileName)?
+                .to_str()
+                .unwrap();
```
Let's use an `except()` for this.

poljar commented on crates/matrix-sdk/src/encryption/futures.rs:
```diff
@@ -0,0 +1,85 @@
+            let mut encryptor = matrix_sdk_base::crypto::AttachmentEncryptor::new(reader);
+
+            let mut buf = Vec::new();
+            encryptor.read_to_end(&mut buf)?;
```
If we end up removing the HTTP trait and only support `reqwests`, can we also get rid of this `read_to_end()` and instead pass in the `Reader` to `reqwests`?

jplatte replied:
No, the only way to not buffer the request body is to put it in a `futures::Stream`, a `Read`er doesn't work. Also the `Stream` thing seems to not be available on WASM (which is why progress tracking also doesn't work on WASM, I should actually `cfg` out those methods).

poljar replied:
Hmm, ok, but I think that we'll have to special case this eventually so we don't use `read_to_end()` at least on non-WASM. Perhaps add an TODO for now, or a GH issue.

jplatte replied:
This is not just for encryption, right? We also need to change Ruma to allow sending media without buffering the whole content in memory. Created https://github.com/matrix-org/matrix-rust-sdk/issues/2106.

poljar replied:
That's correct. Thanks.

poljar commented on crates/matrix-sdk/src/room/joined/futures.rs:
```diff
@@ -0,0 +1,218 @@
+    error::ImageError,
+};
+
+/* pub struct PrepareAndSendAttachment<'a> {
```
What's this about?

jplatte replied:
Ah, this is outdated. Since `prepare_and_send_attachment` is an internal function I ended up adding an extra `send_progress` parameter which is simpler than the builder-style API that `send_attachment()` got.

jplatte commented:
Rebased earlier, but apparently there's new conflicts. Let me resolve those before you re-review..

jplatte commented:
Rebased again. The previous fixup commit is now squashed, but the new one intentionally not. You should only have to review the last three commits, everything else is the same as before modulo merge conflicts.

poljar commented on crates/matrix-sdk/Cargo.toml:
```diff
@@ -16,20 +16,10 @@ features = ["docsrs"]
-    "matrix-sdk-sqlite?/crypto-store",        # activate crypto-store on sqlite if given
-    "matrix-sdk-indexeddb?/e2e-encryption",   # activate on indexeddb if given
-]
+e2e-encryption = ["matrix-sdk-base/e2e-encryption", "matrix-sdk-base/message-ids", "matrix-sdk-sqlite?/crypto-store", "matrix-sdk-indexeddb?/e2e-encryption"]
```
I agree with the other two, but this one seems to be quite long for it to be on a single line.

jplatte replied:
This was not an intentional change. Not idea what happened there..

poljar commented on crates/matrix-sdk/src/client/futures.rs:
```diff
@@ -0,0 +1,99 @@
+    /// Note that any subscribers obtained from
+    /// [`subscribe_to_send_progress`][Self::subscribe_to_send_progress]
+    /// will be invalidated by this.
+    #[cfg_vis(target_arch = "wasm32", pub(crate))]
```
Neat, I need this in vodozemac in one place.
